### ISSUE TITLE
Remote Kernel does not start on Windows 7

### ISSUE THREAD

**Issue by Rudey** [nteract/hydrogen#1398]
**Description:**

I can't start a remote kernel with Hydrogen on my server.
I want to use the atom-editor to write code in Python and Matlab on a client machine (win10 and ubuntu 18.04LTS) and run this code via Hydrogen on a remote machine (win7) on my local network because it has more power...

I configured my Hydrogen gateways as follows:

```
"[
    {
        "name": "Lab basement",
        "options":{
            "baseUrl": "http://localhost:9999",
            "token": "super_secret_token"
        }
    }
]"
```

If I do `Hydrogen: Connect To Remote Kernel` in Atom, I can select the Remote machine and also `'Python 3'` and `'New Session'`, but it does not start up a new python kernel as expected.

**Additional Info:**

I connect to the remote machine via ssh and have my local port 9999 forwarded to 8888 with `ssh -N -L 9999:localhost:8888 <UserName>@<ServerName>`. Furthermore, I start a jupyter-notebook server on the remote machine via ssh and it is hosted on 8888 with the above token (as explained in the Hydrogen Documentation). So when I open `localhost:9999` in my browser and enter the token, I get the jupyter notebook as usual. I can start new projects and kernels, run code etc. as I should. When starting via ssh the root directory of the notebook is within `C:/cygwin64/home/<UserName>`. I also tried starting a jupyter-notebook directly on the server which then runs in the `C:/Users/<UserName>` directory. This is because on Windows 7 I need to use Cygwin to make my SSH server.

**Steps to Reproduce:**

1. Start Jupyter Notebook on remote
2. In Atom select `Hydrogen: Connect To Remote Kernel`
3. Try to start a new session on remote

**Versions:**

```
Atom            : 1.29.0
Electron        : 2.0.5
Chrome          : 61.0.3163.100
Node            : 8.9.3
Jupyter         : 4.4.0
Jupyter Notebook: 5.6.0
```
Client: Windows 10
Server: Windows 7

**Logs:**

The remote juypter notebook console shows the following error:

```
[W 15:09:33.358 NotebookApp] 404 POST /api/sessions?1535465383171 (::1): <PathToUserDir>RemoteKernelTest\test.py-af0f8195-53ac-45b3-9fad-81c6450c20b3 is outside root contents directory
[W 15:09:33.358 NotebookApp] <PathToUserDir>RemoteKernelTest\test.py-af0f8195-53ac-45b3-9fad-81c6450c20b3 is outside root contents directory
[W 15:09:33.358 NotebookApp] 404 POST /api/sessions?1535465383171 (::1) 0.00ms referer=None
```

and the Atom Developer console on the client says:

```
16:09:43.264 <PathToUserDir>\.atom\packages\Hydrogen\node_modules\@jupyterlab\services\lib\session\default.js:621
API request failed: Invalid Status: 404
onSessionError @ <PathToUserDir>\.atom\packages\Hydrogen\node_modules\@jupyterlab\services\lib\session\default.js:621
16:09:43.264 internal/process/next_tick.js:188 Uncaught (in promise)
Objectevent: undefinedmessage:
    "Invalid Status: 404"
    request:
    cache: false
    contentType: "application/json"
    data: "{
        "kernel":{"name":"python3"},
        "path":"<PathToUserDir>\\RemoteKernelTest\\test.py-af0f8195-53ac-45b3-9fad-81c6450c20b3",
        "type":"",
        "name":""
        }"
    method: "POST"
    url: "http://localhost:9999/api/sessions"
    __proto__: Object
    settings:
    baseUrl: "http://localhost:9999"
    password: ""
    requestHeaders: {}
    timeout: 0
    token: "super_secret_token"
    user: ""
    withCredentials: false
    wsFactory: ƒ
    wsFactory(url, protocol)
    wsUrl: "ws://localhost:9999/"
    xhrFactory: ƒ
    xhrFactory()
    __proto__: Object
    xhr: exports.XMLHttpRequest
    DONE: 4
    HEADERS_RECEIVED: 2
    LOADING: 3
    OPENED: 1
    UNSENT: 0
    abort: ƒ ()
    addEventListener: ƒ (event, callback)
    dispatchEvent: ƒ (event)
    getAllResponseHeaders: ƒ ()
    getRequestHeader: ƒ (name)
    getResponseHeader: ƒ (header)
    handleError: ƒ (error)
    onabort: ƒ (event)
    onerror: ƒ (event)
    onload: ƒ (event)
    onreadystatechange: null
    ontimeout: ƒ (event)
    open: ƒ (method, url, async, user, password)
    readyState: 4
    removeEventListener: ƒ (event, callback)
    responseText:
    "{
        "message":
        "<PathToUserDir>\\RemoteKernelTest\\test.py-af0f8195-53ac-45b3-9fad-81c6450c20b3 is outside root contents directory",
        "reason": null
    }"
    responseXML: ""
    send: ƒ (data)
    setDisableHeaderCheck: ƒ (state)
    setRequestHeader: ƒ (header, value)
    status: 404
    statusText: null
    timeout: 0
    withCredentials: false
    __proto__: Object
    __proto__: Object
```

**Possible Explanation**

What I suspect is happening here is that Hydrogen asks the remote server to start a new kernel in the directory of the currently opened python script with a path that is on the client. Of course this does not work because the script does not exist on the server as I am coding on the client. Is there something wrong with my configuration? How can I get this to run?

**Comment by Rudey** [comment]
Btw when I try to run code locally on the server with Hydrogen and Atom it works. With 'locally' I mean that I sit in front of the computer and type stuff in. It is just a powerful computer we use for digital image correlation that has windows only software on it.

**Comment by nikitakit** [comment]
If you open a notebook via the web browser, does hydrogen allow you to connect to that kernel? (When selecting existing kernels, kernels that are associated with notebooks should be listed). If so, can you successfully connect to the pre-existing kernel and run code?

**Comment by Rudey** [comment]
Yes I can connect to a pre-existing kernel, run code, and have access to variables on both ends.

**Comment by nikitakit** [comment]
I've seen this issue before where you can access existing kernels but not create any new ones. I used the workaround of creating all of my kernels via the web ui.

This isn't a Windows-specific issue (I had it happen on a linux-hosted JupyterHub instance).

**Comment by Rudey** [comment]
So I must admit that I do not have a lot of expertise in oop and js but I have been digging into where this wrong path might come from.

I also found that I can not choose the right kernels on the client. So when I want to start a new session I can only choose the kernels that are installed on the client (Python3) but the server also supports Matlab kernels, which is available to choose when I work on the server locally.

Here is what I found out so far:

The 'wrong' path is generated in line 93 in ws-kernel-picker.js in the WSKernelPicker class:
```
this._path = `${store.filePath || "unsaved"}-${v4()}`;
```
because `store.filePath` is the local path of the file on the client.

So when selecting `[New Session]` the value of `this._path` is then written into a `_.map` within `WSKernelPicker.onSession`. To 'fix' this I hardcoded a valid path on my server into line 337 in ws-kernel-picker:

```
path: `${"C:\\cygwin64\\home\\<user-name>\\" || "unsaved"}-${v4()}`
```

This starts a new session and gives it the path name as kernel name. You can also then restart and shutdown the kernel with hydrogen. You still can only choose kernels that are available on the client because with this 'fix' I just inject the path before the WSKernel Object is sent to the server.

I do not know whether there is a general fix for this. You would either need to get a valid path from the server somehow, or add a config option for this.
